corehq/apps/domain/tests/test_auth_decorators.py

```diff
@@ -0,0 +1,38 @@
+from django.test import SimpleTestCase
+
+
+CHECK_FAILED = 'check failed'
+
+
+def challenge(check):
+    def decorator(view):
+        def wrapper(request, *args, **kwargs):
+            auth = check(request)
+            if auth:
+                return view(request, *args, **kwargs)
+
+            return CHECK_FAILED
+        return wrapper
+    return decorator
+
+
+passing_decorator = challenge(lambda request: True)
+failing_decorator = challenge(lambda request: False)
+
+
+class LoginOrChallengeTest(SimpleTestCase):
+
+    def test_challenge(self):
+        request = object()
+
+        @passing_decorator
+        def test(request, *args, **kwargs):
+            return 'it worked!'
+
+        self.assertEqual('it worked!', test(request))
```

Comment:
I don't think so since it's the decorator's responsibility to handle that and the test is just a mock. really i'm not sure this test is actually that useful, I just wrote it as I was wrapping my head around what `_login_or_challenge` was actually doing and figuring out the primitives I'd need to test it.